File: semgrep_asvs/rules/custom/javascript/constant-time-compare.js
const crypto = require('crypto');
function verify(k, m, sig) {
  // ruleid: constant-time-compare
  if (crypto.createHmac('sha256', k).update(m).digest('hex') === sig) return true;
  // ruleid: constant-time-compare
  if (sig != crypto.createHmac('sha256', k).update(m).digest('hex')) return false;
  // ok: constant-time-compare
  const expected = Buffer.from(crypto.createHmac('sha256', k).update(m).digest('hex'));
  return crypto.timingSafeEqual(expected, Buffer.from(sig));
}

File: semgrep_asvs/rules/custom/python/constant-time-compare.py
import hmac


def verify(k, m, sig):
    # ruleid: constant-time-compare
    if hmac.new(k, m, "sha256").hexdigest() == sig:
        return True
    # ruleid: constant-time-compare
    if sig != hmac.new(k, m, "sha256").digest():
        return False
    # ok: constant-time-compare
    return hmac.compare_digest(hmac.new(k, m, "sha256").hexdigest(), sig)
